xwiki-platform-core/xwiki-platform-ratings/xwiki-platform-ratings-api/src/main/java/org/xwiki/ratings/internal/averagerating/XObjectAverageRatingManager.java

```diff
@@ -115,8 +124,84 @@ public long removeAverageRatings(EntityReference entityReference) throws Ratings
+            // sub-elements of the page.
+            for (BaseObject ratingsObject : actualDoc
+                .getXObjects(AverageRatingClassDocumentInitializer.AVERAGE_RATINGS_CLASSREFERENCE)) {
+                moveAverageRatingsObject(oldReference, newReference, ratingsObject, changedAverageRatings);
+            }
+
+            // Save the document if something has changed
+            if (!changedAverageRatings.isEmpty()) {
+                XWikiContext context = this.contextProvider.get();
+                this.getObservationManager()
+                    .notify(new UpdatingAverageRatingEvent(), this.getIdentifier(), changedAverageRatings);
+                try {
+                    context.getWiki().saveDocument(actualDoc, "Move average ratings XObjects", true, context);
+                    this.getObservationManager()
+                        .notify(new UpdatedAverageRatingEvent(), this.getIdentifier(), changedAverageRatings);
+                } catch (XWikiException e) {
+                    this.getObservationManager()
+                        .notify(new UpdateAverageRatingFailedEvent(), this.getIdentifier(), changedAverageRatings);
+                }
+            }
+        } catch (Exception e) {
+            throw new RatingsException(
+                String.format("Error while moving the average ratings from [%s] to [%s]", oldReference, newReference),
+                e);
+        }
+
+        return changedAverageRatings.size();
+    }
+
+    private void moveAverageRatingsObject(EntityReference oldDocumentEntityReference,
+        EntityReference newDocumentEntityReference,
+        BaseObject ratingsObject, List<AverageRating> changedAverageRatings)
+    {
+        String xobjectManagerId = ratingsObject.getStringValue(AverageRatingQueryField.MANAGER_ID.getFieldName());
+
+        // Checks that the object is attached to the current manager
+        // Checks that the xobject references the old entity or one of its children.
+        if (Objects.equals(getIdentifier(), xobjectManagerId)) {
+            String reference = ratingsObject.getStringValue(AverageRatingQueryField.ENTITY_REFERENCE.getFieldName());
+            EntityReference oldEntityReference =
+                this.entityReferenceConverter.convert(EntityReference.class, reference);
+            if (oldEntityReference.equals(oldDocumentEntityReference)
+                || oldEntityReference.hasParent(oldDocumentEntityReference))
+            {
+                // If the entity reference of the average ratings object is the same as the old document reference, we
+                // use the new document reference.
+                // If the entity reference a child of the old entity reference, we  update its parent document with the
+                // new document reference.
+                EntityReference newEntityReference;
+                if (oldEntityReference.equals(oldDocumentEntityReference)) {
```

Comment:
you perform twice the same check. Sounds like the code would be simplified with something like: 
```
EntityReference newEntityReference = null;
if (oldEntityReference.equals(oldDocumentEntityReference)) {
  newEntityReference = newDocumentEntityReference;
} else if (oldEntityReference.hasParent(oldDocumentEntityReference)) {
  newEntityReference = oldEntityReference.replaceParent(newDocumentEntityReference);
}
if (newEntityReference != null) {
 // perform the update.
}
```

Reply:
imo this is breaking the separation of concern, the first if checks if we need the handle the object, the second one perform a transformation. 
But I'm fine switch to your version.

Reply:
It's more a question of style again I think, you can keep it like that if you prefer.

Reply:
Too late ;) I opted for the least amount of LoC.